Added support for custom uploading URL

buh opened the pull request:
Add `Uploader` protocol for custom uploading.

VojtaStavik commented on Sources/Client/Client/Client+Request.swift:
```diff
@@ -68,6 +68,36 @@ extension Client {
     func request<T: Decodable>(endpoint: Endpoint,
                                _ progress: @escaping Progress,
                                _ completion: @escaping Completion<T>) -> Cancellable {
+        if let uploader = uploader, let completion = completion as? Completion<FileUploadResponse> {
```
Why doesn't `Client` conform to the `Uploader` protocol too? I think it would make the logic simpler and the solution more flexible. 
Or even better! We can extract the logic from `Client` into the `DefaultUploader` which will be used unless you change it to custom one. Then the `Client` wouldn't need to worry at all if it uses the default or a custom uploader.

buh replied:
I updated PR and please check my message in Slack.

VojtaStavik commented on Sources/Core/Presenter/UploadingItem.swift:
```diff
@@ -156,71 +135,44 @@ public final class UploaderItem: Equatable {
-        
+    private func encodeData() {
         if type == .file || type == .video {
             if let url = url, let data = try? Data(contentsOf: url) {
```
The error is not handled correctly here. We shouldn't ignore the error from `Data(contentsOf:)` and use it on L144.

VojtaStavik commented on Sources/Core/Presenter/UploadingItem.swift:
```diff
@@ -156,71 +135,44 @@ public final class UploaderItem: Equatable {
-                
-                return .error(ClientError.emptyBody(description: errorDescription))
-            }
+            return
```
Is this `return` correct? If so, we should probably rewrite the whole `if` into `guard`.

VojtaStavik commented on Sources/UI/Chat/Composer/ComposerView.swift:
```diff
@@ -259,8 +259,8 @@ public extension ComposerView {
-        uploader?.reset()
-        imageUploaderItems = []
+        uploadManager?.reset()
+        imageUploadingItems = []
```
What's the difference between `imageUploadingItems` and `UploadManager.items`?

buh replied:
That's the question. I don't remember why, but I'll check now.

buh replied:
Fixed/Refactored.

buh commented on Sources/Client/Client/Client+Channel.swift:
```diff
@@ -478,42 +478,42 @@ public extension Client {
-        sendFile(endpoint: .sendImage(fileName, mimeType, imageData, channel), progress, completion)
+                   channel: Channel,
+                   progress: @escaping Client.Progress,
+                   completion: @escaping Client.Completion<URL>) -> Cancellable {
```
No worries about breaking changes.
1. We already have breaking changes in the upcoming release.
2. I can say 99% don't use these methods directly. We've just introduced LLC but as I know all our customers except one use all our frameworks and the default behavior.
3. I think it's good changes and the `Uploader` protocol looks more solid.

buh commented on Sources/Client/Client/Client+Channel.swift:
```diff
@@ -530,9 +530,7 @@ public extension Client {
-    func deleteImage(url: URL,
-                     from channel: Channel,
-                     _ completion: @escaping Client.Completion<EmptyData> = { _ in }) -> Cancellable {
+    func deleteImage(url: URL, channel: Channel, _ completion: @escaping Client.Completion<EmptyData> = { _ in }) -> Cancellable {
```
really don't want to make it an option, because it will lead us to make common Client request completion block optional (`Client+Request.swift` line 55):
```swift
func request<T: Decodable>(endpoint: Endpoint, _ completion: @escaping Completion<T>) -> Cancellable
```
In 1.x I mapped `EmptyData` (because we need codable empty data type) to `Void`, but I think we should avoid side effects and be clear what is happening.

VojtaStavik replied:
- What's wrong about having the completion optional everywhere, when it's actually really optional?
- I don't understand how's `EmptyData` related to this.

buh replied:
If we want an optional completion block we should make in the general request (`Client+Request.swift`) is optional for all requests. We can change it ofc.

VojtaStavik replied:
OK, let's start talking about it and keep this as is for now.

buh commented on Sources/Client/Client/Client+Config.swift:
```diff
@@ -55,10 +55,11 @@ extension Client {
         ///     - stayConnectedInBackground: when the app will go to the background,
-        ///                                  start a background task to stay connected for 5 min
+        ///                                  start a background task to stay connected for 5 min.
+        ///     - uploading: a custom flow for uploading with a custom URL.
```
Old stuff.

buh commented on Sources/Client/Client/Client+Request.swift:
```diff
@@ -108,7 +110,7 @@ extension Client {
             performInCallbackQueue { completion(.failure(error)) }
         } catch {
-            completion(.failure(.unexpectedError(description: error.localizedDescription, error: error)))
+            performInCallbackQueue { completion(.failure(.unexpectedError(description: error.localizedDescription, error: error))) }
```
This is some kind of _bug_ that was missed.
But this case is unexpectable because of all errors from try/catch block can throw a `CliientError`.

buh commented on Sources/Client/Client/Endpoint.swift:
```diff
@@ -54,9 +54,9 @@ public enum Endpoint {
+    case sendImage(Data, _ fileName: String, _ mimeType: String, Channel)
     /// Upload a file to a channel.
-    case sendFile(_ fileName: String, _ mimeType: String, Data, Channel)
+    case sendFile(Data, _ fileName: String, _ mimeType: String, Channel)
```
The same as the above. We should make `Endpoint` internal because we provide API for endpoints with a specific callback type without this type the request with some endpoint useless, you need to know response type.

buh commented:
@VojtaStavik I know I changed and cleaned up these uploading requests to make it more solid to avoid breaking changes in the future. Because only now we introduce the way to work with it and we have the first customer for that. In review try to think about how it will look in the future.

buh commented on Sources/Client/Extensions/Codable+Extensions.swift:
```diff
@@ -166,8 +166,8 @@ extension Encodable {
+/// Empty `Codable` data.
 public struct EmptyData: Codable {
-    /// A shared instance of empty data.
+    /// A shared instance of empty `Codable` data.
```
1. I made it class.
2. I found one place where it was used as `EmptyData()` in `Endpoint`.

VojtaStavik replied:
But why do we need a shared instance of `EmptyData`?

buh replied:
I missed understanding you. Ofc it's better to leave it as a struct.
So, do you prefer to use `EmptyData()`?

buh replied:
In this case, we should make a public `init`, because by default will be available only init with a decoder. So, I think it's handier to have `.empty`

VojtaStavik replied:
I wasn't suggesting something :) I really just asked why we need a shared instance of `EmptyData`.

buh replied:
It needs external frameworks like Core.

VojtaStavik replied:
I think we don't understand each other. Let's leave it as it, it's just a detail.

buh replied:
It's actually not shared. A type for empty decodable data.

buh commented on CHANGELOG.md:
```diff
@@ -24,7 +47,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `MessageViewStyle.pointedCornerRadius` to make pointed corner rounded [#191](https://github.com/GetStream/stream-chat-swift/issues/191). 
-- Added methods for `AvatarView` customization [#203](https://github.com/GetStream/stream-chat-swift/issues/203):
+- Added methods for `AvatarView` customization [#203](https://github.com/GetStream/stream-chat-swift/issues/203).
+- Added `Uploader` protocol. Use them to create own uploader for your file storage. Assign your uploader into `ChannelPresenter` [#100](https://github.com/GetStream/stream-chat-swift/issues/100):
```
I think it was after rebase. Fixed.

buh commented on Sources/Core/Presenter/UploadingItem.swift:
```diff
@@ -1,45 +1,17 @@
-
 /// An uploader item.
-public final class UploaderItem: Equatable {
+public final class UploadingItem: Equatable {
```
It's like _internal_ type. We need to make it public to available for UI framework. All breaking changes won't affect customers and again we will make a release with breaking changes.
I agree and we will do it in this way in the future. If you insist I can add `typealias`.

buh commented on Sources/Core/Presenter/UploadManager.swift:
```diff
@@ -0,0 +1,126 @@
+    ///
+    /// - Parameter item: an uploading item.
+    public func upload(item: UploadingItem) {
+        items.insert(item, at: 0)
```
1. For UI it needs to make it in different orders.
2. A good point. I renamed it as `add(..)` to the manager.

buh commented on Sources/Core/Presenter/UploadManager.swift:
```diff
@@ -0,0 +1,126 @@
+            return .error(error)
+        }
+        
+        // Return shared uploading.
```
We use `UICollectionView` for uploading images. When the cell is dequeued with the same uploading item and subscribed for the uploading progress we should return created shared uploading progress events.

VojtaStavik replied:
Oh, I meant to expand the comment, not to put it here :) I'll be more clear next time.

buh commented on Sources/Core/Presenter/UploadManager.swift:
```diff
@@ -0,0 +1,126 @@
+        items = []
+    }
+    
+    public func startUploading(item: UploadingItem) -> Observable<ProgressResponse<URL>> {
```
It was inside `UploadingItem` I refactored and moved the uploading logic to the manager from the uploading model.

VojtaStavik commented on Sources/Core/Presenter/Channel/ChannelPresenter.swift:
```diff
@@ -169,7 +169,11 @@ extension ChannelPresenter {
+        
+        var attachments = uploadManager.images.isEmpty
+            ? uploadManager.files.compactMap({ $0.attachment })
+            : uploadManager.images.compactMap({ $0.attachment })
```
Is it OK it's not possible to have a message with both images and files?

buh replied:
It should be work separately.
I thought about why it's so easy to make a mistake to work with a single list of items where it's always needed two of them. Each list of items has own presentation and management.

VojtaStavik commented on Sources/Core/Presenter/UploadManager.swift:
```diff
@@ -26,21 +28,38 @@ public final class UploadManager {
+            images.insert(item, at: 0)
+        } else {
+            files.insert(item, at: 0)
+        }
```
I think using `switch` would be better here. We'll get support from the compiler when adding a new type.

VojtaStavik commented on Sources/Core/Presenter/UploadManager.swift:
```diff
@@ -121,5 +141,7 @@ public final class UploadManager {
             })
             .share()
+        
+        return item.uploading ?? .empty()
```
What would cause `item.uploading` being `nil` here?

buh replied:
That's a good question. `UploadingItem` is creating without uploading a subscription. Then `UploadManager` called for `startUploading(item:)` it will set a shared observable for the item as `.uploading`. Another reason, that a model could be created with an attachment and no need to upload it.

VojtaStavik replied:
I referred to this very line (L145). You assign a the `uploading` value on L117 and then on L145 it can be `nil` again and you use ` ?? .empty()` as a fallback value. I think it will never be `nil`. That's why I asked about why it can be `nil`, becase the code suggests it's valid for it to be `nil`.

buh replied:
@VojtaStavik  I think this is a question of coding style. In this specific case, I need to set a value to an optional property and at the same time return it as a non-optional value. So, I was lazy, I know that I can return `.empty()`. In other cases probably I won't have it and I need to make a local var and return it. So, what do you think?

VojtaStavik replied:
We don't need to complicate it and use just `return item.uploading!`. We're 100% sure the value is there (we create it on the previous line). If `item.uploading` was `nil` here, it would mean our assumptions are terribly wrong and we should know about it anyway :)

buh replied:
>  If item.uploading was nil here, it would mean our assumptions are terribly wrong and we should know about it anyway :)
The problem is that our customers will know about that first. The usual case in our tests and Example app everything is working as expected, but our customers could start complaining that on production with millions of people something wrong. It would be really bad if it will be a crash.

VojtaStavik replied:
We're both talking hypothetical situations so it doesn't matter. Leave it as is.

VojtaStavik commented on Sources/UI/Chat/Composer/ComposerView.swift:
```diff
@@ -275,7 +273,8 @@ public extension ComposerView {
     internal func updateSendButton() {
-        let isAnyFileUploaded = uploadManager?.items.first(where: { $0.attachment != nil }) != nil
+        let isAnyFileUploaded = uploadManager?.images.first(where: { $0.attachment != nil }) != nil
+            || uploadManager?.files.first(where: { $0.attachment != nil }) != nil
```
😱 this expression is massive. Mind splitting it into something more readable?

buh replied:
Done. I really tried :)

VojtaStavik commented on Sources/UI/Chat/Composer/ComposerView.swift:
```diff
@@ -273,11 +273,11 @@ public extension ComposerView {
         
         if let style = style {
-            let isHidden = text.isEmpty && !isAnyFileUploaded
+            let isHidden = text.isEmpty && anyUploadedItem == nil
```
What about something like this:
```
        var hasUploadedItem = false
        if let uploadManager = uploadManager {
            let hasUploadedImages = uploadManager.images.contains(where: { $0.attachment != nil })
            let hasUploadedFiles = uploadManager.files.contains(where: { $0.attachment != nil })
            hasUploadedItem = hasUploadedImages || hasUploadedFiles
        }

        if let style = style {
            let isHidden = text.isEmpty && hasUploadedItem
            
```

buh replied:
I think in my case is a pattern to find any uploaded item in 2 lines and the second line is optional if the first one will find an item.
In your case, 6 lines and all of them will be executed.
As a compromise, I think I can add a comment for easier reading.

VojtaStavik replied:
I don't think we're in the fewer-lines-win competition here. What's easier to read and understand is what matters. @b-onc we need a third opinion here.

b-onc replied:
Not because it's fewer lines, but because second expression doesn't need to be run, and I didn't feel like it's hard to read, I'd suggest current way with a minor change

```swift
guard let uploadManager = uploadManager else {
  // no need to check for style of text
  sendButton.isHidden = false
  return
}

let hasUploadedItem = uploadManager.contains(where: {...}) || uploadManager.contains(where: {...})
```

or what @VojtaStavik suggested without `hasUploadedImages` and `hasUploadedFiles` variables (so || expressions directly)

VojtaStavik replied:
@b-onc 
- like the `guard` in the beginning 👍 
- with writing it in one expression you're just doing compiler's job. in my original code, the compiler optimizes it such that if `hasUploadedImages` was true,  `hasUploadedFiles` would never be evaluated, too. Let the compiler do its job. Your job is to write it so your teammates don't have to think about it ;-) 
- we're talking about optimizing code which runs just once and has no impact unless there will be millions of items there (not possible)

b-onc replied:
@VojtaStavik Are you sure the compiler optimizes the code? When you write the expressions explicitly it doesn't optimize them, do you have any resources so I can read up on? I've tried:
```swift
func one() -> Bool {
    let res = Bool.random()
    print("one: \(res)")
    return res
}

func two() -> Bool {
    let res = Bool.random()
    print("two: \(res)")
    return res
}

let oneRes = one()
let twoRes = two()

let val = oneRes || twoRes

if val {
    print("val true")
} else {
    print("val false")
}
```
and when `one()` returns `true`, `two()` is still executed.

If that's the case and compiler optimizes the code, totally agreed that we can have them explicit.

Performance is one thing, and while I do agree that our gains are very minimal here, I don't think readability is taking a hit. I feel writing two expressions explicitly is a bit -unnecessarily- more verbose.

VojtaStavik replied:
You know what, you're probably right! I didn't think about the side effects. Sorry. I will be more careful about my statements.

@buh Pick the solution you like the most and let's merge this!

buh replied:
First of all, I like this conversation because we can figure each other.
1. Yes, I tried to avoid the side effect.
2. I don't think we need a guard for `uploadManager`. It's because we no need it if the user wants to send a message. Some of our customers disable all attachments (basically removed ⊕) and we can optimize that and skip setting upload manager for `ComposerView`, so it will be always `nil`. But I added a guard for the style.
3. I don't want to rename `anyUploadedItem` to `hasUploadedItem` because it will increase complexity as it was at the beginning in one line:
```swift
let x = (a... ?? b...) != nil
```

Can I merge now?

VojtaStavik replied:
OK 👍
